Limiting results returned from SearchRuns API

mparkhe opened the pull request:
- 'max_results' argument in SearchRuns PB
- support in backend stores
- client API changes

Commit:
Limiting results returned from SearchRuns API

- 'max_results' argument in SearchRuns PB
- support in backend stores
- client API changes

Commit:
ordering change: start_time with run_uuid as tie-breaker

Commit:
lint

sueann reviewed (commented):
didn't read all the tests super carefully but wanted to leave the one comment

smurching commented on mlflow/store/abstract_store.py:
```diff
@@ -197,7 +198,7 @@ def get_metric_history(self, run_uuid, metric_key):
 
     @abstractmethod
-    def search_runs(self, experiment_ids, search_filter, run_view_type):
+    def search_runs(self, experiment_ids, search_filter, run_view_type, max_results):
```
Can we make this an optional argument (with default of `SEARCH_MAX_RESULTS_DEFAULT`) to avoid breaking the AbstractStore search_runs API?

smurching commented on mlflow/store/abstract_store.py:
```diff
@@ -206,6 +207,7 @@ def search_runs(self, experiment_ids, search_filter, run_view_type):
         :param search_filter: :py:class`mlflow.utils.search_utils.SearchFilter` object to encode
             search expression or filter string
         :param run_view_type: ACTIVE, DELETED, or ALL runs
+        :param max_results: Maximum number of runs desired
```
Would be nice to add an analogous doc to `list_run_infos` (state that we'll restrict the number of returned runs), unless we plan to remove that API entirely

smurching commented on mlflow/store/sqlalchemy_store.py:
```diff
@@ -399,12 +400,20 @@ def set_tag(self, run_uuid, tag):
+    def search_runs(self, experiment_ids, search_filter, run_view_type, max_results):
+        # TODO: push search query into backend database layer
+        if max_results > SEARCH_MAX_RESULTS_THRESHOLD:
+            raise MlflowException("Search API called with a large max_results ({}). Max threshold "
```
Wording nit/suggestion: `"Invalid value for request parameter max_results. max_results must be at most {}, but got value {}".format(SEARCH_MAX_RESULTS_THRESHOLD, max_results)`

smurching commented on tests/store/test_sqlalchemy_store.py:
```diff
@@ -293,15 +294,16 @@ def test_run_info(self):
 
-    def _get_run_configs(self, name='test', experiment_id=None, tags=(), parent_run_id=None):
+    def _get_run_configs(self, name='test', experiment_id=None,
+                         tags=(), parent_run_id=None, start_time=int(time.time())):
```
Unfortunately in Python setting `start_time=int(time.time())` will cause the default value to be initialized once & reused across function calls, instead of computed with each function call, so I'd undo this change.

mparkhe replied:
Scary stuff... 

Lemme default this to `None` and call `time.time()` inside the function, if not supplied by caller.

smurching reviewed (commented):
Had some small comments, but otherwise LGTM, thanks @mparkhe!

mparkhe commented on mlflow/protos/service.proto:
```diff
@@ -832,6 +832,9 @@ message SearchRuns {
   // Defaults to only active runs.
   optional ViewType run_view_type = 3 [default = ACTIVE_ONLY];
 
+  // Maximum number of runs desired. Max threshold is 50000
```
Whoa! Good catch.

Commit:
addressing review comments
- default argument values for max_results in stores
- fixed confusing language
- max threshold set to 50K
- worked around python default argument idiosyncrasies.

mparkhe commented:
@smurching, @sueann : addressed your feedback, in case you want to take another look at this.

sueann commented on mlflow/store/abstract_store.py:
```diff
@@ -207,7 +208,7 @@ def search_runs(self, experiment_ids, search_filter, run_view_type, max_results)
             search expression or filter string
         :param run_view_type: ACTIVE, DELETED, or ALL runs
-        :param max_results: Maximum number of runs desired
+        :param max_results: Maximum number of runs desired. Default: 1000
```
Putting the default here (and in other doc strings) seems brittle since it may change, though it is nice to have it for documentation. Not sure if there is a better way though?

mparkhe replied:
Good one. We could reference the constant, using `py:const` but it displays the constant name and `mlflow.store` is not documented.

![image](https://user-images.githubusercontent.com/6900999/56236906-945f7200-603f-11e9-8481-ed000382f192.png)

Also, looks like default arguments are pre-filled with `1000`. So I could go ahead and remove it from the `:param` line in docs.

sueann reviewed (commented):
just one comment/question - lgtm otherwise.

Commit:
fixed docs around "max_results" default argument value
